File: _Api/Feed/FeedService.cs
using MightyRSS._Api.Feed.Types;
using MightyRSS.Data.Records;
using MightyRSS.Data.UoW;
using System;
using System.Linq;
using System.Net;
using WJBCommon.Lib.Api.Type;
using WJBCommon.Lib.Data;

namespace MightyRSS._Api.Feed
{
    public interface IFeedService
    {
        Result<AddFeedSourceResponse> AddFeedSource(UserRecord user, AddFeedSourceRequest request);
        Result<GetFeedResponse> GetFeed(UserRecord user);
        Result DeleteFeedSource(UserRecord user, Guid reference);
        Result AddFeedToCollection(UserRecord user, Guid feedReference, AddFeedToCollectionRequest request);
    }

    public sealed class FeedService: IFeedService
    {
        private readonly IUnitOfWorkFactory<IMightyUnitOfWork> _mightyUnitOfWorkFactory;
        private readonly IFeedReaderService _feedReaderService;

        public FeedService(
            IUnitOfWorkFactory<IMightyUnitOfWork> mightyUnitOfWorkFactory,
            IFeedReaderService feedReaderService)
        {
            _mightyUnitOfWorkFactory = mightyUnitOfWorkFactory;
            _feedReaderService = feedReaderService;
        }

        public Result<AddFeedSourceResponse> AddFeedSource(UserRecord user, AddFeedSourceRequest request)
        {
            using var unitOfWork = _mightyUnitOfWorkFactory.Create();

            var feedSource = unitOfWork.FeedSources.GetByRssUrl(request.Url);
            if (feedSource == null)
            {
                var feedDetailsResult = _feedReaderService.Read(request.Url, null);
                if (feedDetailsResult.IsFailure)
                    return Result<AddFeedSourceResponse>.Error(feedDetailsResult.ErrorMessage);

                var feedDetails = feedDetailsResult.Value;

                feedSource = new FeedSourceRecord
                {
                    Reference = feedDetails.Reference,
                    Title = feedDetails.Title,
                    Description = feedDetails.Description,
                    RssUrl = feedDetails.RssUrl,
                    WebsiteUrl = feedDetails.WebsiteUrl,
                    Articles = feedDetails.Articles
                        .Select(x => new FeedSourceRecord.Article
                        {
                            Url = x.Url,
                            Title = x.Title,
                            Summary = x.Summary,
                            PublishedAt = x.PublishedAt,
                            PublishedAtAsString = x.PublishedAtAsString,
                            Author = x.Author
                        })
                        .ToList(),
                    ArticlesUpdatedAt = DateTime.Now.ToLocalTime()
                };

                unitOfWork.FeedSources.Save(feedSource);
            }

            var userFeedSource = new UserDataFeedSourceRecord
            {
                User = user,
                FeedSource = feedSource,
                Collection = null
            };

            unitOfWork.UserFeedSources.Save(userFeedSource);

            unitOfWork.Commit();

            return Result<AddFeedSourceResponse>.Of(new AddFeedSourceResponse
            {
                Reference = feedSource.Reference,
                Title = feedSource.Title,
                Description = feedSource.Description,
                RssUrl = feedSource.RssUrl,
                WebsiteUrl = feedSource.WebsiteUrl,
                Collection = userFeedSource.Collection,
                Articles = feedSource.Articles.ConvertAll(x => new AddFeedSourceResponse.FeedArticle
                {
                    Url = x.Url,
                    Title = x.Title,
                    Summary = x.Summary,
                    Author = x.Author,
                    PublishedAt = x.PublishedAt,
                    PublishedAtAsString = x.PublishedAtAsString
                })
            });
        }

        public Result<GetFeedResponse> GetFeed(UserRecord user)
        {
            using var unitOfWork = _mightyUnitOfWorkFactory.Create();

            var feedSources = unitOfWork.UserFeedSources.GetFeedSources(user);

            unitOfWork.Commit();

            return Result<GetFeedResponse>.Of(new GetFeedResponse
            {
                Sources = feedSources.ConvertAll(source => new GetFeedResponse.FeedSource
                {
                    Reference = source.FeedSource.Reference,
                    Title = source.FeedSource.Title,
                    Description = source.FeedSource.Description,
                    RssUrl = source.FeedSource.RssUrl,
                    WebsiteUrl = source.FeedSource.WebsiteUrl,
                    Collection = source.Collection,
                    Articles = source.FeedSource.Articles.ConvertAll(article => new GetFeedResponse.FeedArticle
                    {
                        Url = article.Url,
                        Title = article.Title,
                        Summary = article.Summary,
                        Author = article.Author,
                        PublishedAt = article.PublishedAt,
                        PublishedAtAsString = article.PublishedAtAsString
                    })
                })
            });
        }

        public Result DeleteFeedSource(UserRecord user, Guid reference)
        {
            using var unitOfWork = _mightyUnitOfWorkFactory.Create();

            var userFeedSource = unitOfWork.UserFeedSources.GetByUserAndFeedSourceReference(user, reference);
            if (userFeedSource == null)
                return Result.Error("Sorry, unable to delete feed source as it could not be found.");

            unitOfWork.UserFeedSources.Delete(userFeedSource);

            unitOfWork.Commit();

            return Result.Success(HttpStatusCode.NoContent);
        }

        public Result AddFeedToCollection(UserRecord user, Guid feedReference, AddFeedToCollectionRequest request)
        {
            var unitOfWork = _mightyUnitOfWorkFactory.Create();

            var userFeedSource = unitOfWork.UserFeedSources.GetByUserAndFeedSourceReference(user, feedReference);
            if (userFeedSource == null)
                return Result.Error("The feed source with the given reference could not be found for you.", HttpStatusCode.NotFound);

            userFeedSource.Collection = request.Collection;

            unitOfWork.UserFeedSources.Update(userFeedSource);

            unitOfWork.Commit();

            return Result.Success();
        }
    }
}

File: frontend/src/service/Feed.service.ts
import dayjs from 'dayjs';

import { feedApi } from '@/api/feed/Feed.api';

import { AddFeedSourceRequest } from '@/api/feed/types/AddFeedSource.type';
import { IFeedArticle } from '@/model/FeedArticle.type';

class FeedService {

    async getFeed(): Promise<Array<IFeedArticle> | Error> {
        const getFeedResponse = await feedApi.getFeed();

        if (getFeedResponse instanceof Error)
            return getFeedResponse;

        return getFeedResponse.sources.map(source => source.articles
            .map(article => ({
                url: article.url,
                title: article.title,
                summary: article.summary,
                author: article.author,
                publishedAt: dayjs(article.publishedAt ?? article.publishedAtAsString),
                source: {
                    reference: source.reference,
                    title: source.title,
                    description: source.description,
                    rssUrl: source.rssUrl,
                    websiteUrl: source.websiteUrl,
                    collection: source.collection,
                    titleAlias: source.titleAlias,
                },
            })))
            .flat();
    }

    async addFeedSource(request: AddFeedSourceRequest): Promise<Array<IFeedArticle> | Error> {
        const addFeedSourceResponse = await feedApi.addFeedSource(request);

        if (addFeedSourceResponse instanceof Error)
            return addFeedSourceResponse;

        return addFeedSourceResponse.articles.map(x => ({
            url: x.url,
            title: x.title,
            summary: x.summary,
            author: x.author,
            publishedAt: dayjs(x.publishedAt ?? x.publishedAtAsString),
            source: {
                reference: addFeedSourceResponse.reference,
                title: addFeedSourceResponse.title,
                description: addFeedSourceResponse.description,
                rssUrl: addFeedSourceResponse.rssUrl,
                websiteUrl: addFeedSourceResponse.websiteUrl,
                collection: addFeedSourceResponse.collection,
                titleAlias: null,
            },
        }));
    }

    async deleteFeedSource(reference: string): Promise<void | Error> {
        const deleteFeedSourceResponse = await feedApi.deleteFeedSource(reference);

        if (deleteFeedSourceResponse instanceof Error)
            return deleteFeedSourceResponse;
    }
}

export const feedService = new FeedService();